hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java

```diff
@@ -0,0 +1,223 @@
+       * -> Deleted after the old snapshot was taken, which means it will be
+       *    present in oldKeyTable and missing in newKeyTable.
+       * -> Modified after the old snapshot was taken, which means it will be
+       *    present in oldKeyTable and present in newKeyTable with same
+       *    Object ID but with different metadata.
+       * -> Renamed after the old snapshot was taken, which means it will be
+       *    present in oldKeyTable and present in newKeyTable but with different
+       *    name and same Object ID.
+       */
+
+      final String oldKeyName = oldObjIdToKeyMap.get(id);
+      final String newKeyName = newObjIdToKeyMap.get(id);
+
+      if (oldKeyName == null && newKeyName == null) {
+        // This cannot happen.
+        continue;
+      }
+
+      // Key Created.
+      if (oldKeyName == null) {
+        createDiffs.add(DiffReportEntry.of(DiffType.CREATE, newKeyName));
+        continue;
+      }
+
+      // Key Deleted.
+      if(newKeyName == null) {
+        deleteDiffs.add(DiffReportEntry.of(DiffType.DELETE, oldKeyName));
+        continue;
+      }
+
+      // Key modified.
+      if(oldKeyName.equals(newKeyName)) {
+        modifyDiffs.add(DiffReportEntry.of(DiffType.MODIFY, newKeyName));
+        continue;
+      }
+
+      // Key Renamed.
+      renameDiffs.add(DiffReportEntry.of(DiffType.RENAME,
+          oldKeyName, newKeyName));
+    }
+    /*
+     * The order in which snap-diff should be applied
+     *
+     *     1. Delete diffs
+     *     2. Rename diffs
+     *     3. Create diffs
+     *     4. Modified diffs
+     *
+     * Consider the following scenario
+     *
+     *    1. File "A" is created.
+     *    2. File "B" is created.
+     *    3. File "C" is created.
+     *    Snapshot "1" is taken.
+     *
+     * Case 1:
+     *   1. File "A" is deleted.
+     *   2. File "B" is renamed to "A".
+     *   Snapshot "2" is taken.
+     *
+     *   Snapshot diff should be applied in the following order:
+     *    1. Delete "A"
+     *    2. Rename "B" to "A"
+     *
+     *
+     * Case 2:
+     *    1. File "B" is renamed to "C".
+     *    2. File "B" is created.
+     *    Snapshot "2" is taken.
+     *
+     *   Snapshot diff should be applied in the following order:
+     *    1. Rename "B" to "C"
+     *    2. Create "B"
+     *
+     */
+
+    final List<DiffReportEntry> snapshotDiffs = new ArrayList<>();
+    snapshotDiffs.addAll(deleteDiffs);
+    snapshotDiffs.addAll(renameDiffs);
+    snapshotDiffs.addAll(createDiffs);
+    snapshotDiffs.addAll(modifyDiffs);
+    return snapshotDiffs;
+  }
+
+  private BucketLayout getBucketLayout(final String volume,
+                                       final String bucket,
+                                       final OMMetadataManager mManager)
+      throws IOException {
+    final String bucketTableKey = mManager.getBucketKey(volume, bucket);
+    return mManager.getBucketTable().get(bucketTableKey).getBucketLayout();
```

Comment:
Are `mManager.getBucketKey(volume, bucket)` and `mManager.getBucketTable().get(bucketTableKey)` nullable?

Reply:
`mManager.getBucketKey(volume, bucket)` can never return null.
`mManager.getBucketTable().get(bucketTableKey)` should not return null once we implement proper locking logic in the snapshot code. For now, it is theoretically possible to get a null pointer exception here.

Reply:
Thanks for clarification.